### PR TITLE
Update record stacks to build assoc stacks using associated config

### PR DESCRIPTION
**Pull request by jcredding**
This updates record stacks to build their records associations
record stacks using a factory config for the association. This is
done by getting a configured factory config for an association.
This makes it so a factory config only needs to know about its
records associations and doesn't need to copy factories from other
factory configs. Since the copying logic is no longer needed this
removes it. This is also setup for a future effort to simplify the
association factory keys (they no longer need to store the owner
record class since they don't copy other owners factories).

This updates the `Factory::Config` to have a `factory_config_for`
method that takes a record association, finds a matching factory
for it and returns the factories config. This builds on the
existing `factories_for` and `factory_for` logic.

Next, this updates record stacks to get an associated factory
config and use that when building a record stack for an
association. This makes it so all of the record stacks that
are built as associations are walked don't use the same factory
config which means the factory config doesn't need to know
about configured factories for deeply nested associations.

This also updates the `Factory::Config` to remove the copy
factories from logic since it's no longer needed. This makes the
configs store less information and do less work when adding an
association factory.

@kellyredding - Ready for review. This doesn't significantly affect performance in our large app's test suite (good or bad) but it does simplify the logic and will make it simpler to add the feature for having record stacks force walk associations.

### CODE REVIEW

File: test/unit/factory/record_stack_tests.rb [hunk 1]
```diff
@@ -408,7 +412,7 @@ class AssociationDataTests < UnitTests
 
     should "return its preset when getting record data for a preset association" do
       preset_record = @associated_record_class.new
-      @association.writer(preset_record)
+      @association.owner.send("#{@association.reflection.name}=", preset_record)
```
Comment (kellyredding) [inline test/unit/factory/record_stack_tests.rb]: @jcredding why this change?

Comment (jcredding) [inline test/unit/factory/record_stack_tests.rb]: @kellyredding - Ahh yes, I messed up in #262 when I changed this to `writer`. Fake record associations don't have a `writer` method they have a `write` method this is confusing because activerecord associations have a `writer` method. I got confused and used the wrong one (and forgot to run tests) and then when I saw the test failing in this effort I was annoyed and just used `send` to avoid the whole issue.

I wasn't originally aware of the `writer` method on activerecord associations but I noticed it while working on MR this round. I have a note to sync them up to avoid confusion even though I don't currently have any "requirement" that they be the same.
